x-pack/plugins/lens/public/xy_visualization/to_expression.ts

```diff
@@ -26,8 +26,8 @@ export const toExpression = (
   state.layers.forEach((layer) => {
     metadata[layer.layerId] = {};
     const datasource = datasourceLayers[layer.layerId];
-    datasource.getTableSpec().forEach((column) => {
-      const operation = datasourceLayers[layer.layerId].getOperationForColumnId(column.columnId);
+    datasource?.getTableSpec().forEach((column) => {
```

Comment:
Why are these changes necessary? I see how they are fixing the functional test failure, but it's very concerning this is necessary. `toExpression` should never be called without consistent datasource layers. I can't figure out what change in your PR caused that to happen.